worker/src/sidechain_storage/storage.rs

```diff
@@ -0,0 +1,887 @@
+		Ok(SidechainStorage { db, shards, last_blocks })
+	}
+
+	/// gets all shards of currently loaded sidechain db
+	pub fn shards(&self) -> &Vec<ShardIdentifier> {
+		&self.shards
+	}
+
+	/// gets the last block of the current sidechain DB and the given shard
+	pub fn last_block_of_shard(&self, shard: &ShardIdentifier) -> Option<&LastSidechainBlock> {
+		self.last_blocks.get(shard)
+	}
+
+	/// gets the block hash of the sidechain block of the given shard and block number, if there is such a block
+	pub fn get_block_hash(
+		&self,
+		shard: &ShardIdentifier,
+		block_number: BlockNumber,
+	) -> Result<Option<BlockHash>> {
+		match self.db.get((*shard, block_number).encode()).map_err(Error::OperationalError)? {
+			None => Ok(None),
+			Some(enocded_hash) => Ok(Some(
+				BlockHash::decode(&mut enocded_hash.as_slice()).map_err(Error::DecodeError)?,
+			)),
+		}
+	}
+
+	/// gets the block of the given blockhash, if there is such a block
+	#[allow(unused)]
+	pub fn get_block(&self, block_hash: &BlockHash) -> Result<Option<SignedSidechainBlock>> {
+		match self.db.get(block_hash.encode()).map_err(Error::OperationalError)? {
+			None => Ok(None),
+			Some(enocded_hash) => Ok(Some(
+				SignedSidechainBlock::decode(&mut enocded_hash.as_slice())
+					.map_err(Error::DecodeError)?,
+			)),
+		}
+	}
+
+	/// update sidechain storage
+	pub fn store_blocks(&mut self, blocks_to_store: Vec<SignedSidechainBlock>) -> Result<()> {
+		let mut batch = WriteBatch::default();
+		let mut new_shard = false;
+		for signed_block in blocks_to_store.into_iter() {
+			// check if current block is the next in line
+			let current_shard = signed_block.block().shard_id();
+			let current_block_nr = signed_block.block().block_number();
+			if self.shards.contains(&current_shard) {
+				if let Some(last_block) = self.last_block_of_shard(&current_shard) {
+					if last_block.number != current_block_nr - 1 {
+						error!("[Sidechain DB] Sidechainblock (nr: {:?}) is not a succession of the previous block (nr: {:?}) in shard: {:?}",
+						current_block_nr, last_block.number, current_shard);
+						continue
+					}
+				} else {
+					error!(
+						"[Sidechain DB] Shard {:?} does not have a last block. Skipping block (nr: {:?}) inclusion",
+						current_shard, current_block_nr
+					);
+					continue
+				}
+			} else {
+				self.shards.push(current_shard);
+				new_shard = true;
+			}
+			// Block hash -> Signed Block
+			let current_block_hash = signed_block.hash();
+			batch.put(&current_block_hash.encode(), &signed_block.encode().as_slice());
+			// (Shard, Block number) -> Blockhash (for block pruning)
+			batch.put(
+				&(current_shard, current_block_nr).encode().as_slice(),
+				&current_block_hash.encode(),
+			);
+			// (last_block_key, shard) -> (Blockhash, BlockNr) current blockchain state
+			let current_last_block =
+				LastSidechainBlock { hash: current_block_hash, number: current_block_nr };
+			self.last_blocks.insert(current_shard, current_last_block.clone());
```

Comment:
This way, when storing a new block, no file read is necessary. The last block can just be looked up from the in memory hashmap and be used to control if the next block is really a valid successor in terms of blocknumber. 
I think also later on, when a different worker wants to know if it's out of date, it can just peak into the memory, no file access.

But I have not performance tested it in anyway. Might as well be an overkill.

Reply:
This will depend on the db size. It might actually be relevant, if we don't have a cache layer otherwise. So, fine to keep it, but maybe extract it to separate function to ensure that those two are always set together?

Reply:
I've added an `add_block` function, which adds a block to the `WriteBatch`. I think a completely separate function does not make sense here, because a block should always be added to the whole bunch of DB, not just appended to the "LastBlock"

I'll add a unit test for this function tomorrow, just let me know if that approach statisfies you?